File: src/main/java/com/matag/game/cardinstance/ability/selector/PowerToughnessConstraintUtils.java
package com.matag.game.cardinstance.ability.selector;

import com.matag.cards.ability.selector.PowerToughnessConstraint;
import com.matag.game.cardinstance.CardInstance;

public class PowerToughnessConstraintUtils {
  public static boolean check(PowerToughnessConstraint powerToughnessConstraint, CardInstance cardInstance) {
    var value = getValue(powerToughnessConstraint.getPowerOrToughness(), cardInstance);

    return switch (powerToughnessConstraint.getPowerToughnessConstraintType()) {
      case EQUAL -> value == powerToughnessConstraint.getValue();
      case LESS -> value < powerToughnessConstraint.getValue();
      case LESS_OR_EQUAL -> value <= powerToughnessConstraint.getValue();
      case GREATER -> value > powerToughnessConstraint.getValue();
      case GREATER_OR_EQUAL -> value >= powerToughnessConstraint.getValue();
    };
  }

  private static int getValue(PowerToughnessConstraint.PowerOrToughness powerOrToughness, CardInstance cardInstance) {
    return switch (powerOrToughness) {
      case POWER -> cardInstance.getPower();
      case TOUGHNESS -> cardInstance.getToughness();
    };
  }
}

File: src/main/kotlin/com/matag/game/cardinstance/ability/selector/PowerToughnessConstraintUtils.kt
package com.matag.cards.ability.selector

import com.matag.cards.ability.selector.PowerToughnessConstraint.PowerOrToughness
import com.matag.game.cardinstance.CardInstance

object PowerToughnessConstraintUtils {
    @JvmStatic
    fun check(powerToughnessConstraint: PowerToughnessConstraint, cardInstance: CardInstance): Boolean {
        val value = getValue(powerToughnessConstraint.powerOrToughness, cardInstance)

        return when (powerToughnessConstraint.powerToughnessConstraintType) {
            PowerToughnessConstraintType.EQUAL -> value == powerToughnessConstraint.value
            PowerToughnessConstraintType.LESS -> value < powerToughnessConstraint.value
            PowerToughnessConstraintType.LESS_OR_EQUAL -> value <= powerToughnessConstraint.value
            PowerToughnessConstraintType.GREATER -> value > powerToughnessConstraint.value
            PowerToughnessConstraintType.GREATER_OR_EQUAL -> value >= powerToughnessConstraint.value
        }
    }

    private fun getValue(powerOrToughness: PowerOrToughness, cardInstance: CardInstance): Int {
        return when (powerOrToughness) {
            PowerOrToughness.POWER -> cardInstance.power
            PowerOrToughness.TOUGHNESS -> cardInstance.toughness
        }
    }
}
